src/snowflake/snowpark/dataframe_transform_functions.py

```diff
@@ -120,3 +164,135 @@ def moving_agg(
+                H: Hours, D: Days, W: Weeks, M: Months, Y: Years. For future-oriented analysis, use positive numbers,
+                and for past-oriented analysis, use negative numbers.
+            sliding_interval: Interval at which the window slides, specified in the same format as the windows.
+                H: Hours, D: Days, W: Weeks.
+            group_by: A list of column names on which the DataFrame is partitioned for separate window calculations.
+            col_formatter: An optional function for formatting output column names, defaulting to the format '<input_col>_<agg>_<window>'.
+                        This function takes three arguments: 'input_col' (str) for the column name, 'operation' (str) for the applied operation,
+                        and 'value' (int) for the window size, and returns a formatted string for the column name.
+
+        Returns:
+            A Snowflake DataFrame with additional columns corresponding to each specified time window aggregation.
+
+        Raises:
+            ValueError: If an unsupported value is specified in arguments.
+            TypeError: If an unsupported type is specified in arguments.
+            SnowparkSQLException: If an unsupported aggregration is specified.
+
+        Example:
+            aggregated_df = df.transform.time_series_agg(
+                time_col='ORDERTIME',
+                group_by=['PRODUCTKEY'],
+                aggs={
+                    'SALESAMOUNT': ['SUM', 'MIN', 'MAX']
+                },
+                sliding_interval='12H',
+                windows=['7D', '14D', '-7D', '-14D', '1T']
+            )
+        """
+        self._validate_aggs_argument(aggs)
+        self._validate_string_list_argument(group_by, "group_by")
+        self._validate_formatter_argument(col_formatter)
+
+        if not windows:
+            raise ValueError("windows must not be empty")
+
+        if not sliding_interval:
+            raise ValueError("sliding_interval must not be empty")
+
+        if not time_col or not isinstance(time_col, str):
+            raise ValueError("time_col must be a string")
+
+        # check for valid time_col names
+
+        slide_duration, slide_unit = self._validate_and_extract_time_unit(
+            sliding_interval, "sliding_interval", allow_negative=False
+        )
+        sliding_point_col = "sliding_point"
+
+        agg_df = self._df
+        agg_df = agg_df.withColumn(
+            sliding_point_col,
+            self._get_sliding_interval_start(time_col, slide_unit, slide_duration),
+        )
+        agg_df.show()
+        agg_exprs = []
+        for column, functions in aggs.items():
+            for function in functions:
+                agg_exprs.append((column, function))
+
+        # Perform the aggregation
+        sliding_windows_df = agg_df.groupBy(group_by + [sliding_point_col]).agg(
+            agg_exprs
+        )
+
+        for window in windows:
```

Comment:
have you checked the generated SQL and what's the performance? We have another alternative to conduct time-series calculation in our user guide: https://docs.snowflake.com/LIMITEDACCESS/snowflake-feature-store#time-series-features. It's pure SQL. But my gut feeling is it will be faster than how it works here.

Reply:
Yes. I have seen that. Both are same. The SQL is creating CTE to create and join the column. The Python code here also is doing the same.